File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/math/P384_ShuffleAnArray.py
import random
from typing import List


class Solution:

    def __init__(self, nums: List[int]):
        self.nums = nums
        self.original = nums.copy()

    def reset(self) -> List[int]:
        return self.original

    def shuffle(self) -> List[int]:
        self.nums = self.original.copy()
        n = len(self.nums)
        for i in range(n):
            r = random.randrange(i, n)
            self.nums[i], self.nums[r] = self.nums[r], self.nums[i]
        return self.nums

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/math/P384_ShuffleAnArray.java
package com.dsa.leetcode.labuladong.math;

import java.util.Arrays;
import java.util.Random;

/**
 * @since 2024/7/27 09:30
 */
public class P384_ShuffleAnArray {

    class Solution {
        int[] nums;
        Random random;

        public Solution(int[] nums) {
            this.nums = nums;
            random = new Random();
        }

        public int[] reset() {
            return nums;
        }

        public int[] shuffle() {
            int n = nums.length;
            int[] copy = Arrays.copyOf(nums, n);
            for (int i = 0; i < n; i++) {
                int r = i + random.nextInt(n - i);
                swap(copy, i, r);
            }
            return copy;
        }

        public void swap(int[] nums, int p, int q) {
            int temp = nums[p];
            nums[p] = nums[q];
            nums[q] = temp;
        }
    }

}

